### PR TITLE
Add generation and chat scripts support for XLA devices

### PR DESCRIPTION
**Pull request by gkroiz**
Added code specific for `xla` devices which is used when Fabric's device type is `xla`

### CODE REVIEW

File: generate.py [hunk 1]
```diff
@@ -101,6 +110,7 @@ def main(
         config = Config(**json.load(fp))
 
     fabric = L.Fabric(devices=1)
+    use_xla = True if fabric.device.type == "xla" else False
```
Comment (carmocca) [inline generate.py:113]:
```suggestion
    use_xla = fabric.device.type == "xla"
```

File: generate_adapter.py [hunk 2]
```diff
@@ -46,6 +46,7 @@ def main(
     check_valid_checkpoint_dir(checkpoint_dir)
 
     fabric = L.Fabric(devices=1)
+    use_xla = True if fabric.device.type == "xla" else False
```
Comment (carmocca) [inline generate_adapter.py:49]:
```suggestion
    use_xla = fabric.device.type == "xla"
```

File: generate.py [hunk 3]
```diff
@@ -42,6 +44,10 @@ def generate(
     empty[:T] = idx
     idx = empty
 
+    if use_xla:
```
Comment (carmocca) [inline generate.py:47]: Should we just check `idx.device.type` so that we don't need to pass an argument around?

**Review by carmocca — COMMENTED** [review]
Can you also update chat.py?

File: chat.py [hunk 4]
```diff
@@ -36,6 +36,10 @@ def generate(
     assert max_seq_length > T
     buffer = max((len(tokens) for tokens in stop_tokens), default=0)
 
+    if model.device.type == "xla":
+        import torch_xla.core.xla_model as xm
+        xm.mark_step()
```
Comment (carmocca) [inline chat.py:41]: What's the reasoning behind calling this before forward is called (before the loop)?

**Review by carmocca — COMMENTED** [review]
Changes look good. Thoughts on this @lantiga?

File: chat.py (hunk 4 above)
Comment (gkroiz) [reply chat.py]: Thanks for pointing this out, I don't think this is needed. I will leave the import there but will remove the mark_step().

Comment (carmocca) [reply chat.py]: @gkroiz imports are cached in Python, it can be inside the loop

**Review by carmocca — APPROVED** [review]

File: generate.py [hunk 5]
```diff
@@ -42,6 +42,10 @@ def generate(
     empty[:T] = idx
     idx = empty
 
+    if idx.device.type == "xla":
```
Comment (carmocca) [inline generate.py:45]: Don't forget about this file! :P

**Review by lantiga — APPROVED** [review]
Looks good. I wonder if at some point we should expose an opt-in Fabric method marking the end of the step for the executors that make use of it.

File: generate.py (hunk 5 above)
Comment (gkroiz) [reply generate.py:45]: I think that even initializing a model will create xla ops, so we should add `xm.mark_step()` back to `chat.py`

**Pull request merged**